wallet-new/src/Cardano/Wallet/API/V1/Errors.hs

```diff
@@ -1,19 +1,104 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
 
 module Cardano.Wallet.API.V1.Errors where
 
 import           Universum
 
-import           Cardano.Wallet.API.V1.Types (WalletError (..))
-import           Data.Aeson (encode)
+import           Data.Aeson
+import           Data.Aeson.TH (deriveJSON)
 import qualified Network.HTTP.Types.Header as HTTP
 import           Servant
+import           Test.QuickCheck (Arbitrary (..), oneof)
 
+import           Cardano.Wallet.API.V1.TH (conNamesList, deriveWalletErrorJSON)
+
+--
+-- Error handling
+--
+
+-- | Type representing any error which might be thrown by wallet.
+--
+-- Errors are represented in JSON in the following way:
+-- ```
+-- {
+--     "message" : <constr_name>,
+--     "diagnostic" : <data>
+-- }
+-- ```
+-- where `<constr_name>` is a string containing name of error's constructor (e. g. `NotEnoughMoney`),
+-- and `<data>` is an object containing additional error data.
+-- Additional data contains constructor fields, field names are record field names without
+-- a `we` prefix, e. g. for `OutputIsRedeem` error "diagnostic" field will be the following:
+-- ```
+-- {
+--     "address" : <address>
+-- }
+-- ```
+-- If constructor does not define record field names, then data values are labeled with
+-- keys "v1", "v2" and so on, e. g. for `NotRecordError` error:
```

Comment:
Uhm, this is a slightly unconventional choice which might create problems to frontend. Cannot we simply forbid constructors without records during TH compilation and fail with an error like "You are required to have a record for each `WalletError` constructor", or something like that?

Reply:
Yes, that's perfectly possible (and will even make code shorter).